src/main/scala/org/downtowndailybread/request/ClientRequest.scala

```diff
@@ -0,0 +1,141 @@
+package org.downtowndailybread.request
+
+import java.sql.Connection
+import java.util.UUID
+
+import org.downtowndailybread.exceptions.client.ClientNotFoundException
+import org.downtowndailybread.exceptions.clientattributetype.ClientAttributeTypeNotFoundException
+import org.downtowndailybread.model.helper.AttribNameValuePair
+import org.downtowndailybread.model.{Client, ClientAttribute, ClientAttributeTypeInternal}
+import spray.json._
+
+class ClientRequest(val conn: Connection) extends DatabaseRequest {
+
+  val clientAttributeTypesRequester = new ClientAttributeTypeRequest(conn)
+
+  def getAllClients(): Seq[Client] = {
+    val sql =
+      s"""
+         |select * from (select distinct on (cid.id,
+         |                                  cattribtype.id) cid.id as userId,
+         |                                  cattribtype.id         as attribTypeId,
+         |                                  cattrib.value,
+         |                                  meta.is_valid
+         |               from canonical_id cid
+         |                           left join canonical_type ctype on cid.type = ctype.id
+         |                           left join client_attribute cattrib on cid.id = cattrib.client_id
+         |                           left join metadata meta on cattrib.metadata_id = meta.rid
+         |                           left join client_attribute_type cattribtype on cattrib.type_id = cattribtype.id
+         |               where 1 = 1
+         |                 and ctype.type = 'client'
+         |               order by cid.id, cattribtype.id, cattrib.rid desc) all_attribs
+         |where is_valid
+      """.stripMargin
+    val statement = conn.prepareStatement(sql)
+
+    val result = statement.executeQuery()
+
+    val clientAttributeTypes = (new ClientAttributeTypeRequest(conn)).getClientAttributeTypes()
+
+    val clients = createSeq(result, result => {
+      val attributeTypeName = result.getString("attribTypeId")
+      clientAttributeTypes.find(_.id.toString == attributeTypeName) match {
+        case Some(attrib) =>
+          Client(
+            UUID.fromString(result.getString("userId")),
+            Seq(ClientAttribute(attrib.tpe, s"""${result.getString("value")}""".parseJson))
+          )
+        case None => // This code SHOULD be unreachable in practice, but in theory we could see it.
+          throw new ClientAttributeTypeNotFoundException(attributeTypeName)
+      }
+    }).groupBy(r => r.id).map { case (k, v) => Client(k, v.flatMap(_.attributes)) }
+
+    clients.toSeq
+  }
+
+  def getClientById(id: UUID): Client = {
+    getAllClients().find(_.id == id) match {
```

Comment:
We should change this. There is going to be a decent amount of work to get all clients. Why not just pass that ID here and let the query in `getAllClients` take care of it? Otherwise, we have to pull all those fields and parse all that json each time.

Reply:
Yeah, I was being lazy because for now it was faster to code, but youre right that we should do it right hte first time because we're probably not going to fix it in the future.